src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs

```diff
@@ -0,0 +1,348 @@
+            if (_modelToDetector.TryGetValue(semanticModel, out var detector))
+            {
+                return detector;
+            }
+
+            return _modelToDetector.GetValue(
+                semanticModel, _ => TryCreate(semanticModel, language));
+        }
+
+        private static RegexPatternDetector TryCreate(
+            SemanticModel semanticModel, RegexEmbeddedLanguage language)
+        {
+            var regexType = semanticModel.Compilation.GetTypeByMetadataName(typeof(Regex).FullName);
+            if (regexType == null)
+            {
+                return null;
+            }
+
+            var methodNamesOfInterest = GetMethodNamesOfInterest(regexType, language.SyntaxFacts);
+            return new RegexPatternDetector(
+                semanticModel, language, regexType, methodNamesOfInterest);
+        }
+
+        public static bool IsDefinitelyNotPattern(SyntaxToken token, ISyntaxFactsService syntaxFacts)
+        {
+            // We only support string literals passed in arguments to something.
+            // In the future we could support any string literal, as long as it has
+            // some marker (like a comment on it) stating it's a regex.
+            if (!syntaxFacts.IsStringLiteral(token))
+            {
+                return true;
+            }
+
+            if (!IsMethodOrConstructorArgument(token, syntaxFacts) && 
+                !HasRegexLanguageComment(token, syntaxFacts, out _))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static bool HasRegexLanguageComment(
+            SyntaxToken token, ISyntaxFactsService syntaxFacts, out RegexOptions options)
+        {
+            if (HasRegexLanguageComment(token.GetPreviousToken().TrailingTrivia, syntaxFacts, out options))
+            {
+                return true;
+            }
+
+            for (var node = token.Parent; node != null; node = node.Parent)
+            {
+                if (HasRegexLanguageComment(node.GetLeadingTrivia(), syntaxFacts, out options))
+                {
+                    return true;
+                }
+            }
+
+            options = default;
+            return false;
+        }
+
+        private static bool HasRegexLanguageComment(
+            SyntaxTriviaList list, ISyntaxFactsService syntaxFacts, out RegexOptions options)
+        {
+            foreach (var trivia in list)
+            {
+                if (HasRegexLanguageComment(trivia, syntaxFacts, out options))
+                {
+                    return true;
+                }
+            }
+
+            options = default;
+            return false;
+        }
+
+        private static bool HasRegexLanguageComment(
+            SyntaxTrivia trivia, ISyntaxFactsService syntaxFacts, out RegexOptions options)
+        {
+            if (syntaxFacts.IsRegularComment(trivia))
+            {
+                var text = trivia.ToString();
```

Comment:
>ToString [](start = 34, length = 8)

I'm concerned by this (I think it's allocating). Is there a way to scan the trivia without allocating?

Reply:
Added comment.